lib/assert.js

```diff
@@ -931,6 +932,79 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+const mustCallChecks = [];
+
+var attachExitHandler = 0;
+
+function runCallChecks(exitCode) {
+  const console = require('console');
+
+  const failed = mustCallChecks.filter((context) => {
+    context.messageSegment = `exactly ${context.exact}`;
+    return context.actual !== context.exact;
+  });
+
+  failed.forEach((context) => {
+    console.error('Mismatched %s function calls. Expected %s, actual %d.',
+                  context.name,
+                  context.messageSegment,
+                  context.actual);
+    console.error(context.stack.split('\n').slice(2).join('\n'));
+  });
+
+  if (exitCode !== 0) return;
+  if (failed.length) process.exitCode = 1;
+}
+
+assert.mustCall = function mustCall(fn, exact = 1) {
+  if (process._exiting)
+    throw new ERR_MUSTCALL_UNAVAILABLE_DURING_EXIT();
+  if (typeof fn === 'number') {
+    exact = fn;
+    fn = noop;
+  } else if (fn === undefined) {
+    fn = noop;
+  }
+
+  if (typeof exact !== 'number')
+    throw new ERR_INVALID_ARG_TYPE(`Invalid exact value: ${exact}`);
+
+  // eslint-disable-next-line no-restricted-syntax
+  const stackTrace = inspect(new Error());
+
+  const context = {
+    exact: exact,
+    actual: 0,
+    stack: stackTrace,
+    name: fn.name || '<anonymous>'
+  };
+
+  // Add the exit listener only once to avoid listener leak warnings
+  if (attachExitHandler === 0) {
+    attachExitHandler = 1;
+    process.on('exit', runCallChecks);
+  }
+
+  mustCallChecks.push(context);
+  return function() {
+    // Increment call count
+    context.actual++;
+    if (context.actual === context.exact) {
+      // Once function has reached its call count remove it from
+      // MustCallChecks list to prevent memory leaks
+      const index = mustCallChecks.indexOf(context, 0);
+      mustCallChecks.splice(index, 1);
+    }
+    if (context.actual > context.exact) {
+      // If call count exceeds exact throw an error
+      throw new ERR_MUSTCALL_CALL_COUNT_EXCEEDED();
```

Comment:
It would be good to include the actual stack trace and the expected number of calls. Otherwise it's difficult to map this back to the original call.

Reply:
The current implementation is already better! Now it'll just be quite verbose and first log a message, than an error with the stacktrace and then the actual error.

Ideally, this would be an `AssertionError` only. All errors thrown by assert should be an instance of that class.

I wonder though if it's ideal to throw on the very first call that is above the expected or to throw on process exit. That way it would be known how many times the function was actually called.

Reply:
Yes, I was unsure to throw immediately after the call count has exceeded the expected number or wait until process exit. Throwing immediately is causing mustCall() functions that would be called normally to throw. I'd like to hear what everyone else thinks

Reply:
I would still throw immediately. If you really want to avoid the other mustCalls from being reported, you can set it to short circuit and disable the exit handler when you check.